storage: use lease start time to determine follower read safety

ajwerner opened the pull request:
This change obviously needs some testing. I'm putting it out to make sure I
understand the issue and proposed solution.

This change attempts to address a situation where a lease transfer may prevent
a follower read. This happens because after the lease transfer we no longer have
any information in the closedts storage regarding the range from the new
leaseholder.

Fixes #35129.
Fixes #32980.

Release note: None

Commit:
storage: use lease start time to determine follower read safety

This change recognizes that lease transfer start times have provide write
invariants equivalent to those provided by the closed timestamp and MLAI and
thus the maximum closed timestamp for a replica can be viewed as the maximum
timestamp of that of the current replica lease start time and the timestamp
provided by the closed timestamp provider. Utilizing this fact makes follower
reads as well as rangefeeds more robust to lease transfers. Before this change
lease transfers to different nodes might cause closed timestamp information for
a range to become unavailable leading to an inability for a range to serve a
read at a timestamp which was previously determined to be safe.

Fixes #35129.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35130)
<!-- Reviewable:end -->

tbg commented on pkg/storage/replica_follower_read.go:
```diff
@@ -66,6 +66,12 @@ func (r *Replica) canServeFollowerRead(
 			}
+			// We may be able to serve the read based on the lease start time.
+			// It is safe to serve a read for requests which occur before this
+			// Replica's view of the lease start time.
```
> This is because anything proposed by a previous leaseholder and yet unobserved won't apply under this new lease, and under this lease nothing can be proposed that would affect timestamps earlier than the start time of the lease (in practice, this is enforced by the timestamp cache on the leaseholder).

The one concern that I have (which is probably not a real one but needs to be sleuthed) is that of lease extensions and how they affect the `Start` timestamp. We basically want `Start` to never move forward during the tenure of a single leaseholder (as measured via the lease sequence number).

It also seems that this works for expiration-based leases as well, so we do seem to have some kind of follower reads for expiration-based ranges, even though it's a pretty shitty one since the Start timestamp won't move forward.

nvb reviewed (commented):
Yeah, this is basically what I had in mind. Thanks for getting the PR out so quickly.

We're also going to want this logic to help keep rangefeed closed timestamps moving forward on lease transfers as well. Here's one proposal to accomplish this and remove duplicated code:
1. get rid of `Provider.CanServe(ts)`, it should be equivalent to `!Provider.MaxClosed.Less(ts)`. Switch over all tests
2. create a new `Replica.MaxClosed` method that takes the closed timestamp and the lease into consideration
3. use this new method in `Replica.canServeFollowerRead` and `Replica.handleClosedTimestampUpdateRaftMuLocked`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35130#-:-LZGW9Q14BaeYLPtTG6z:b8oobok)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @nvanbenschoten, and @tbg)

___
*[pkg/storage/replica_follower_read.go, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35130#-LZGS7Pm-9BoHbQege7v-r1-71:-LZGYdcY-Um6U215buG0:b-ahae0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/184c63c7216325fd0eb8464b78fdb45d21bf4588/pkg/storage/replica_follower_read.go#L71)):*
[This](https://github.com/cockroachdb/cockroach/blob/8b1f7cdcd16b04daa47006ee7ab7fe6089d3922b/pkg/storage/batcheval/cmd_lease_request.go#L59) is where that logic lives. Your intuition holds for everything except for if the min proposed timestamp is set. In that case, we forward the start timestamp to this value even though we're performing an extension. This appears to still work out though, based on the comment:

> // Note also that leasePostApply makes sure to update the timestamp cache in
// this case: even though the lease holder does not change, the the sequence
// number does and this triggers a low water mark bump.

`Lease.Equivalent` will return `false` if the `Start` timestamp of the lease changes, meaning that it will trigger a timestamp cache bump, so the comment checks out.
___
*[pkg/storage/replica_follower_read.go, line 72 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35130#-LZGWP7_A1UE1w2xOcuu:-LZGWP7_A1UE1w2xOcuv:bync85i) ([raw file](https://github.com/cockroachdb/cockroach/blob/184c63c7216325fd0eb8464b78fdb45d21bf4588/pkg/storage/replica_follower_read.go#L72)):*
> ```Go
> 			// It is safe to serve a read for requests which occur before this
> 			// Replica's view of the lease start time.
> 			if ba.Timestamp.Less(lease.Start) {
> ```

I think this should actually be `!lease.Start.Less(ba.Timestamp)`. A leaseholder shouldn't be able to write exactly at its lease's start timestamp.


<!-- Sent from Reviewable.io -->

ajwerner commented:
@nvanbenschoten this still needs some more testing, especially targeted around lease transfers but PTAL and see if this is what you had in mind.

nvb reviewed (commented):
Yes, this is exactly what I had in mind. @tbg does anything stick out to you as obviously wrong about switching from `CanServe` to `MaxClosed`?

Of course, this only makes me more excited for both targetted testing around lease transfers and more general invariant testing with lease movement, splitting, merging, rebalancing, and chaos :smiley: 

> Reviewed 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35130#-:-LZIMceB4oMF-q2imh6v:bai657t)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @tbg)

___
*[pkg/storage/replica_follower_read.go, line 94 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35130#-LZIMIdFAtb24qwydbvA:-LZIMIdFAtb24qwydbvB:b-gdcs7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/f54ba93773053316af3d7b6036bb78be01dc3d4a/pkg/storage/replica_follower_read.go#L94)):*
> ```Go
> 	maxClosed := r.store.cfg.ClosedTimestamp.Provider.MaxClosed(
> 		lease.Replica.NodeID, r.RangeID, ctpb.Epoch(lease.Epoch), ctpb.LAI(lai))
> 	if maxClosed.IsEmpty() || maxClosed.Less(lease.Start) {
> ```

We can replace this block with `maxClosed.Forward(lease.Start)`


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
Nice simplification, I like it.

> Reviewed 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35130#-:-LZJMecK6cUuYFWF7Vft:bh0o34e)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

ajwerner commented:
I had been avoiding pushing an update to this diff because I was writing more tests on splits and merges but those tests have turned out to not be passing and need more investigation so I took the lease related testing from my testing branch and plopped it over here. Should be RFAL.

Commit:
storage: propagate closed timestamp information to RHS after split

This commit prevents discontinuities in ability to read a value from a follower
after a split. The invariant we want to uphold is that values which were safe
to read from a follower will remain safe to read following the split. In the
face of merges the system provides the claim that timestamps which were readable
on both the left and right sides of the merge will remain readable after the
merge.

Release note: None

ajwerner commented:
I added another commit here which makes follower reads robust to splits. I could split it out into a separate PR if that makes this easier to review but the testing relied on code from the original PR.

ajwerner commented:
@nvanbenschoten PTAL when you get a chance.

nvb reviewed (commented):
> Reviewed 4 of 4 files at r3, 6 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35130)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @nvanbenschoten)

___
*[pkg/storage/closed_timestamp_test.go, line 121 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/35130#-L_ifz4CDBbBLNSNPKkS:-L_ifz4CDBbBLNSNPKkT:brpr6wi) ([raw file](https://github.com/cockroachdb/cockroach/blob/b83eb997a32112e76e023cc9c11e953e6cb01e26/pkg/storage/closed_timestamp_test.go#L121)):*
> ```Go
> 	// that we can always read from this timestamp from all replicas even while
> 	// lease transfers are ongoing. The test launches a goroutine to randomly
> 	// trigger transfers every
> ```

Finish this sentence :)
___
*[pkg/storage/closed_timestamp_test.go, line 154 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/35130#-L_igY5A0pxzxMw59bLs:-L_igY5A0pxzxMw59bLt:b-yxyvz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b83eb997a32112e76e023cc9c11e953e6cb01e26/pkg/storage/closed_timestamp_test.go#L154)):*
> ```Go
> 	g.Go(transferLeasesRandomlyUntilDeadline)
> 
> 	// Attempt to send read requests to	a replica in a tight loop until deadline
> ```

Weird spacing.
___
*[pkg/storage/replica.go, line 318 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/35130#-L_imnQv21p7Tx0EDQmK:-L_imnQv21p7Tx0EDQmL:b-zdvu6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ca3ffe2cbd903c50ff65b6a59800021c759116/pkg/storage/replica.go#L318)):*
> ```Go
> 		// initialMaxClosed is the initial maxClosed timestamp for the replica as known
> 		// from its left-hand-side upon creation.
> 		initialMaxClosed hlc.Timestamp
> ```

I'm curious what your proposal was for informing the closed timestamp infrastructure of splits instead of introducing this extra state. It doesn't seem like much of a violation to inform a Provider that all of its knowledge about a (node, rangeID) also applies to a different (node, rangeID).
___
*[pkg/storage/replica_follower_read.go, line 94 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35130#-LZIMIdFAtb24qwydbvA:-L_iivIL3b1wmkFZeAE3:b4cqz0w) ([raw file](https://github.com/cockroachdb/cockroach/blob/f54ba93773053316af3d7b6036bb78be01dc3d4a/pkg/storage/replica_follower_read.go#L94)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We can replace this block with `maxClosed.Forward(lease.Start)`
</blockquote></details>

Looks like this wasn't addressed. We can do the same for `initialMaxClosed` as well.


<!-- Sent from Reviewable.io -->

ajwerner commented:
> I'm curious what your proposal was for informing the closed timestamp infrastructure of splits instead of introducing this extra state. It doesn't seem like much of a violation to inform a Provider that all of its knowledge about a (node, rangeID) also applies to a different (node, rangeID).

It was sort of gross. Modifying the ctpb.Entry values in storage is weird. The basic idea was the clone the MLIA map value in the entries in storage. This feels weird because these are legitimately messages from a remote store that I'll be modifying. That change to the API then has to propagate through the provider. The bigger issue is picking the right MLAI. Before I niavely just cloned what was in there but I don't think that works as there could have been a new update after the split. We sort of need to either have the LAI of the split itself. I don't exactly know where to find that. I'm wavering on whether that's the right path to follow.

nvanbenschoten commented:
> The bigger issue is picking the right MLAI. Before I niavely just cloned what was in there but I don't think that works as there could have been a new update after the split.

I see, we want the largest closed timestamp due to an MLAI at or before the index of the split. Conveniently, the way you have it now gives exactly those semantics. I guess I'm ok with this approach then, although @tbg should also take a look.

Also, make sure you add a nice comment about how this works right at the point where we set `initialMaxClosed`. The dependencies there are subtle. If that method is called with a `r.mu.state.LeaseAppliedIndex > leaseAppliedIndexOfSplit` then we could get an incorrect result and would start serving inconsistent follower reads.

ajwerner reviewed (commented):
Added a comment. Waiting on @tbg

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35130#-:-L_n9v409qLFkrtGTgNN:bz21f4q)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/storage/replica.go, line 318 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/35130#-L_imnQv21p7Tx0EDQmK:-L_n7xxI3wNYweRaAH3B:b-7ecb9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9ca3ffe2cbd903c50ff65b6a59800021c759116/pkg/storage/replica.go#L318)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm curious what your proposal was for informing the closed timestamp infrastructure of splits instead of introducing this extra state. It doesn't seem like much of a violation to inform a Provider that all of its knowledge about a (node, rangeID) also applies to a different (node, rangeID).
</blockquote></details>

Discussed in github. The problems are twofold, 1 is that it's awkward to modify a ctpb.Entry from a different node and creating an interface to do so feels wrong and 2 is that it's hard to know the right MLAI value to put into that entry. Blindly cloning the current value for the LHS might be wrong. I think a reasonable value could be determined but it feels wrong to have to add logic to figure that out.
___
*[pkg/storage/replica_follower_read.go, line 94 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35130#-LZIMIdFAtb24qwydbvA:-L_n8av07jvVEYZs6_SQ:bytckeq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f54ba93773053316af3d7b6036bb78be01dc3d4a/pkg/storage/replica_follower_read.go#L94)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Looks like this wasn't addressed. We can do the same for `initialMaxClosed` as well.
</blockquote></details>

Done. Much cleaner.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> Reviewed 1 of 4 files at r3, 7 of 7 files at r5, 6 of 6 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35130)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

ajwerner commented:
bors r+

ajwerner commented:
bors r+
